src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTextViewRazorDocumentTracker.cs

```diff
@@ -0,0 +1,267 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Projection;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    internal class DefaultTextViewRazorDocumentTracker : RazorDocumentTracker
+    {
+        private readonly DefaultTextViewRazorDocumentTrackerService _service;
+        private readonly ITextView _textView;
+
+        private DocumentId _documentId;
+        private ITextBuffer _razorBuffer;
+        private bool _isSupportedProject;
+        private ITextBuffer _cSharpBuffer;
+        private SourceTextContainer _textContainer;
+        private Workspace _workspace;
+        private WorkspaceRegistration _workspaceRegistration;
+
+        public override event EventHandler ContextChanged;
+
+        public DefaultTextViewRazorDocumentTracker(DefaultTextViewRazorDocumentTrackerService service, ITextView textView)
+        {
+            if (service == null)
+            {
+                throw new ArgumentNullException(nameof(service));
+            }
+
+            if (textView == null)
+            {
+                throw new ArgumentNullException(nameof(textView));
+            }
+
+            _service = service;
+            _textView = textView;
+
+            Update();
+
+            _textView.Closed += TextView_Closed;
+
+            _textView.BufferGraph.GraphBufferContentTypeChanged += BufferGraph_GraphBufferContentTypeChanged;
+            _textView.BufferGraph.GraphBuffersChanged += BufferGraph_GraphBuffersChanged;
+        }
+
+        public override bool IsSupportedDocument => _isSupportedProject && _textContainer != null;
+
+        public override ProjectId ProjectId => _documentId?.ProjectId;
+
+        public override SourceTextContainer TextContainer => _textContainer;
+
+        public override Workspace Workspace => _workspace;
+
+        private bool Update()
```

Comment:
This code is pretty much impossible to test unfortunately. It depends on statics from Roslyn and I don't think it really makes much sense to decouple it.